File: in_Java/0735 Asteroid Collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> st = new Stack<>();
        for(int i : asteroids){
            if(st.isEmpty()) st.push(i);
            else{
                if(st.peek() < 0 && i > 0) st.push(i);
                else if(i < 0 && st.peek() > 0){
                    int n = i;
                    while(!st.isEmpty() && st.peek() > 0 && n < 0){
                        if(n + st.peek() == 0){n = 0; st.pop();}
                        else{
                            n = Math.abs(n) > Math.abs(st.peek())? n : st.peek();
                            st.pop();
                        }
                    }
                    if(n != 0) st.push(n);
                }else{
                    st.push(i);
                }
            }
        }
        int[] res = new int[st.size()];
        int id = res.length-1;
        while(!st.isEmpty()) res[id--] = st.pop();
        return res;
    }
    /*
    private boolean diff(int a, int b){
        if(a > 0 && b < 0) return true;
        if(a < 0 && b > 0) return true;
        return false;
    }
    */
}

File: in_Python/0735 Asteroid Collision.py
class Solution(object):
    def collide(self, a, b):
        if a > 0 and b < 0:
            return True
        return False
    def asteroidCollision(self, asteroids):
        """
        :type asteroids: List[int]
        :rtype: List[int]
        """
        st = []
        st.append(asteroids[0])
        for i in range(1, len(asteroids)):
            num = asteroids[i]
            if len(st) == 0:
                st.append(num)
                continue
            if not self.collide(st[-1], num):
                st.append(num)
            else:
                while num != 0 and len(st) != 0 and self.collide(st[-1], num):
                    p = st.pop()
                    if abs(p) > abs(num):
                        num = p
                    elif abs(p) == abs(num):
                        num = 0
                if num != 0:
                    st.append(num)
        return st